### PR TITLE
[tt-train] Set up nightly Python unit test runs

### PR DESCRIPTION
**Pull request by kevinwuTT**
### Summary
* Add Python unit test option to L2 nightly workflow. 
* Fix failing Python tests and disable one that needs more investigation.

### Changes:

Black reformatted the entirety of the modified files, so some changes will seem irrelevant. 
See this link to view the diff that ignores the refactor commit: https://github.com/tenstorrent/tt-metal/compare/6682ce3e43525b2d7d6d99e1930e55de371abc66...ctr-kevinwu/tt-train-ci-python

**CI Workflows**

**`.github/workflows/tt-metal-l2-nightly.yaml`**
- New workflow_dispatch input: `run_tt_train_python_unit_tests` (default: `false`). Essentially a copy of `tt-train-cpp-unit-tests` but removes the `gtest_filter` parameter and adds the `wheel-artifact-name` parameter. 
  - Follows the same schedule/trigger as the tt-train C++ option and will run on the same architecture matrix.

**`.github/workflows/tt-train-post-commit.yaml`**
- Adds `wheel-artifact-name` optional input. 
  - If this is not empty, then the Python wheel will be installed along with a few extra Python environment variable setups. Then the Python-only tests will run. The C++ tests will be skipped.

**Python/C++ alignment**
- **`ttml/models/linear_regression.py`**: Weights and biases use `ttnn.DataType.BFLOAT16` to match default Tensor Datatype in C++ `LinearLayer`.
- **`nanobind/nb_models.cpp`**: Registers `LinearLayer` in the `linear_regression` submodule for in Nanobind to fix error: "Unable to convert function return value to a Python type!" error.
- **`ttnn/.../nlp_create_qkv_heads_nanobind.cpp`**: Adds `#include <nanobind/stl/tuple.h>` for tuple binding to fix error:  "Unable to convert function return value to a Python type!" error. 

**Tests**
- **`test_ttml.py`** 
  - Use `BFLOAT16` for Tensor Parameters
  - Disable `TestCppOptimizersWithPythonModules::test_sgd_optimizer_with_python_module` with ticket: https://github.com/tenstorrent/tt-metal/issues/40257
- **`test_custom_operations.py`**
  - Use `BFLOAT16` for Tensor Parameters
  - Redundant device fixture removed from `TestCustomOperationsWithDevice`.
- **`test_randn.py`**
  - Redundant device fixture removed from `TestRandn`.
- **`test_custom_operations.py`**, **`test_nanogpt.py`**, **`test_reshape_op.py`**
  - Call `set_requires_grad(True)` on tensors that need gradients.

### Checklist

- [x] [Sanity tests](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ctr-kevinwu/tt-train-ci-python)
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:ctr-kevinwu/tt-train-ci-python)
- [x] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ctr-kevinwu/tt-train-ci-python)
- [x] [clang-tidy code analysis](https://github.com/tenstorrent/tt-metal/actions/workflows/code-analysis.yaml?query=branch:ctr-kevinwu/tt-train-ci-python)

### CODE REVIEW

**Review by vmelnykovTT — APPROVED** [review]

**Review by abustamanteTT — APPROVED** [review]

**Review by zbaczewskiTT — APPROVED** [review]

**Review by imichalakTT — APPROVED** [review]

**Review by dpopovTT — APPROVED** [review]

**Review by jmalone-tt — APPROVED** [review]

File: .github/workflows/tt-metal-l2-nightly.yaml [hunk 1]
```diff
@@ -441,6 +446,24 @@ jobs:
...
+      fail-fast: false
+      matrix:
+        test-group: ${{ fromJson(needs.generate-arch-matrix.outputs.matrix-tt-train-tests) }}
+    uses: ./.github/workflows/tt-train-post-commit.yaml
+    with:
+      arch: ${{ matrix.test-group.arch }}
+      runner-label: ${{ matrix.test-group.runner-label }}
+      docker-image: ${{ needs.build-artifact.outputs.dev-docker-image }}
+      build-artifact-name: ${{ needs.build-artifact.outputs.build-artifact-name }}
+      timeout: 90
```
Comment (williamlyTT) [inline .github/workflows/tt-metal-l2-nightly.yaml]: I don't see the tests taking longer than ~5 minutes, can this be lowered to like 10 minutes max?

**Review by bbradelTT — APPROVED** [review]
